core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/state/ReplicationState.java

```diff
@@ -41,16 +44,47 @@ public ReplicationState(Metadata metadata, Host localhost)
 
     public Map<LongTokenRange, ImmutableSet<Host>> getTokenRangeToReplicas(TableReference tableReference)
     {
+        String keyspace = tableReference.getKeyspace();
+
         Map<LongTokenRange, ImmutableSet<Host>> tokenRangeToReplicaMap = new HashMap<>();
 
-        for (TokenRange tokenRange : myMetadata.getTokenRanges(tableReference.getKeyspace(), myLocalHost))
+        Map<ImmutableSet<Host>, ImmutableSet<Host>> replicaCache = new HashMap<>();
+
+        for (TokenRange tokenRange : myMetadata.getTokenRanges(keyspace, myLocalHost))
         {
             LongTokenRange longTokenRange = convert(tokenRange);
-            ImmutableSet<Host> hosts = ImmutableSet.copyOf(myMetadata.getReplicas(tableReference.getKeyspace(), tokenRange));
-            tokenRangeToReplicaMap.put(longTokenRange, hosts);
+            ImmutableSet<Host> replicas = ImmutableSet.copyOf(myMetadata.getReplicas(keyspace, tokenRange));
+
+            tokenRangeToReplicaMap.put(longTokenRange, getOrCache(replicas, replicaCache));
```

Comment:
This is all a bit tangled.

I'm guessing that we're doing this so that we can avoid to keep multiple identical instances Set<>.

One way to perhaps better convey the intent would be to replace the content of the loop with this (and get rid of the first getOrCache() method):

```
LongTokenRange longTokenRange = convert(tokenRange);
ImmutableSet<Host> replicas = replicaCache.computeIfAbsent(myMetadata.getReplicas(keyspace, tokenRange), hosts -> ImmutableSet.copyOf(hosts));
tokenRangeToReplicaMap.put(longTokenRange, replicas);
```
As a bonus, we'd avoid creating temporary copies as well!

You'll have to change the type of replicaCache to `Map<Set<Host>, ImmutableSet<Host>>`, but that should be OK?

Reply:
Done